content/posts/improving-outbox.md

```diff
@@ -0,0 +1,155 @@
+
+The second issue is the non-deterministic deduplication data eviction. The information about the processed messages is removed based on its age, based on the assumption that duplicates are more likely to be spaced closely on the time axis and that the likelihood of a duplicate message arriving goes down super-linearly with the time distance between the duplicate copies. In other words, the most prevalent strategy is to keep the deduplication data for a week and hope that it is enough. Well, we think we can do better.
+
+## Decoupling
+
+In this post we will attempt to address the first problem. Let's look again at the Outbox pattern
+
+{{< highlight c "linenos=inline,hl_lines=,linenostart=1" >}}
+var entity = await persister.LoadByCorrelationId(correlationId)
+                   ?? new Entity { Id = correlationId };
+
+TransportOperation[] outgoingMessages;
+if (!entity.OutboxState.ContainsKey(messageId))
+{
+    var state = (T)entity.BusinessState ?? new T();
+
+    var (newState, pendingTransportOperations) =
+        await handlerCallback(state, context);
+
+    outgoingMessages = pendingTransportOperations.Operations
+                                                 .Serialize();
+
+    entity.BusinessState = newState;
+    entity.OutboxState[messageId] = new OutboxState
+    {
+        OutgoingMessages = outgoingMessages
+    };
+
+    await persister.Persist(entity);
+}
+else
+{
+    outgoingMessages = entity.OutboxState[messageId]
+                             .OutgoingMessages;
+}
+
+if (outgoingMessages != null)
+{
+    var toDispatch = outgoingMessages.Deserialize();
+    await Dispatch(context, toDispatch);
+
+    entity.OutboxState[messageId].OutgoingMessages = null;
+
+    await persister.Persist(entity);
+}
+{{< / highlight >}}
+
+The root cause of the issue can be easily seen in line 34 where we remove the content of the outgoing messages but we leave the ID of the incoming message. Why? Because we need it for the deduplication check done in line 5. 
+
+What can we do to remove the need for keeping all these message IDs around inside out entity state? Notice the requirement for correct deduplication is that we need to keep the IDs of all processed messages for the entire deduplication period without any interruptions but it does not state where these IDs are stored. As long as they are **in at least one location**, we can move them around at will. So how about this:
+
+{{< highlight c "linenos=inline,hl_lines=,linenostart=1" >}}
+await deduplicationStore.Add(messageId);
+entity.OutboxState.Remove(messageId);
+await persister.Persist(entity);
+{{< / highlight >}}
+
+This code satisfies this requirement as we first add the deduplication information to an external store and then persist the new state of the entity. Now we need to take this new store into account in the deduplication check. Previously we only checked the collection inside the entity
```

Comment:
Should we note that this makes some assumptions on the consistency model of the two storages? We need changes in `deduplicationStore` to be visible before the removal of message-id from the entity state is visible.